File: helpers/graph/cpp/tsort.cpp
#include <stack>
#include <vector>

using namespace std;

// Does not support graphs that have a weight
// on their edges.
struct TopologicalSort {
    bool *s;
    stack<int> rpo;

    TopologicalSort(vector<int> g[], int n)
    {
        s = new bool[n];
        for (int v = 0; v < n; v++)
            if (!s[v]) dfs(g, v, n);
    }

    void dfs(vector<int> g[], int v, int n)
    {
        s[v] = true;
        for (const auto &u : g[v])
            if (!s[u]) dfs(g, u, n);
        rpo.push(v);
    }
};

File: helpers/graph/py/tsort.py
#!/usr/bin/env python3


class Node:
    def __init__(self, value):
        self.value = value
        self.edges = []
    
    def add_edge(self, dst):
        self.edges.append(dst)


def topological_sort(u):
    seen = set()
    stack = []
    def dfs(u):
        seen.add(u)
        for v in u.edges:
            if v not in seen:
                dfs(v)
        stack.append(u)
    dfs(u)
    for v in u.edges:
        if v not in seen:
            dfs(v)
    stack.reverse()
    return stack


a = Node("A")
b = Node("B")
c = Node("C")
d = Node("D")
e = Node("E")
a.add_edge(b)
a.add_edge(c)
b.add_edge(d)
c.add_edge(d)
c.add_edge(e)
print(*map(lambda x: getattr(x, "value"), topological_sort(a)))
